Feature/outlet maker 148

richpsharp opened the pull request:
Creates a `pygeoprocessing.routing.detect_outlets` function that creates a vector with points placed on the center of outlet pixel on 
 a D8 flow direction raster. The vector has additional metadata about the (i, j) locations of the pixel in raster coordinate space which can be useful when turning the vector into a raster or using as some other indexing mechanism at the raster level.

This PR adds that function, has a unit test to test all the boundary conditions of an outlet detection, and makes an update to `HISTORY.rst`.

Closes #148 

p.s. there's one other change where i moved the call to open a vector for rasterization to after the section where the code first checks to see if there's valid input. That's the `vector = gdal.OpenEx(vector_path, gdal.OF_VECTOR)` line you'll see modified in `geoprocessing.py`. It functionally doesn't change anything but conceptually it's a nice idea to not allocate resources before checking to see if the state of the system is valid.

richpsharp committed:
re #148 committing base function

richpsharp committed:
Merge branch 'feature/geometric-stream-extraction-128' into feature/outlet-maker-148

richpsharp committed:
Merge branch 'bugfix/stats-worker-core-dump-146' into feature/outlet-maker-148

richpsharp committed:
re #148 point testing

richpsharp committed:
Merge branch 'main' into feature/outlet-maker-148

richpsharp committed:
re #148 fixing issue where only check the flow direction

richpsharp committed:
fixes #148 adding history and updating test

richpsharp committed:
Merge branch 'feature/geometric-stream-extraction-128' into feature/outlet-maker-148

richpsharp committed:
re #148 merge

richpsharp committed:
re #148 for some reason there was outlets in there

richpsharp committed:
re #148 putting function back

richpsharp committed:
try it with root

richpsharp committed:
re #148 need to offset the projected coordinates too

richpsharp committed:
Merge branch 'main' into feature/outlet-maker-148

richpsharp committed:
re #148 simplifying the is_outlet test

richpsharp committed:
re #148 cleaning up history

emlys commented on src/pygeoprocessing/routing/routing.pyx:
```diff
@@ -4076,6 +4076,111 @@ def calculate_subwatershed_boundary(
+                    outlet_feature = ogr.Feature(outlet_layer.GetLayerDefn())
+                    outlet_feature.SetGeometry(outlet_point)
+                    outlet_feature.SetField('i', xi_root)
+                    outlet_feature.SetField('j', yi_root)
```
I'm wondering about including `i` and `j` as the original raster coordinates, what would be the use case for that? It seems potentially confusing to give each point an extra pair of coordinates. And without additional metadata like the pixel size and original spatial reference, would `i` and `j` lose their meaning?

richpsharp replied:
I mentioned it in the docstring and added a comment in c134127e48167e25e946523b4604cd9e05a50319 explaining this.. these aren't the projected coordinates but the raster i,j coordinates. It's useful for when you're using this datastructure to seed an upstream search from a pixel coordinate rather than taking the step of having the projected coordinate and then inverse geotransforming it back into raster space. All the hydrological algorithms we have operate at the raster/pixel scale so that's just a handy way of getting it back down there.

A very concrete example of this is where I just used it where I needed to calculate the number of beneficiaries "downstream" from a point. You can do this by using a "distance to channel" function with a weighted raster that is the beneficiary. However you need to also provide where your channels are. Since I actually wanted it to go to the outlet I use this data structure to make a raster that only has the pixels set to 1 where these points are.

I see what you mean that it looks extra and could be derived anyway, but the use case for this is in the raster space not the projected space. In that context the extra information is actually the projected coordinates, but it's nice to have those because then you can just plink it down in qgis and everything matches up.

How does that sound?

emlys replied:
Makes sense! thanks for the explanation

emlys commented on src/pygeoprocessing/routing/routing.pyx:
```diff
@@ -4076,6 +4076,111 @@ def calculate_subwatershed_boundary(
+                    outlet_feature.SetGeometry(outlet_point)
+                    outlet_feature.SetField('i', xi_root)
+                    outlet_feature.SetField('j', yi_root)
+                    outlet_layer.CreateFeature(outlet_feature)
```
should there also be a unique id field? Remembering a comment in delineateit that I didn't write: _"It's important to have a user-facing unique ID field for post-processing (e.g. table-joins) that is not the FID. FIDs are not stable across file conversions that users might do."_

richpsharp replied:
Sure, added in c134127e48167e25e946523b4604cd9e05a50319 and added a test for it in 7c6d48beb86f5530211cf739b82866fea8baa37b.

davemfish replied:
> should there also be a unique id field? Remembering a comment in delineateit that I didn't write: _"It's important to have a user-facing unique ID field for post-processing (e.g. table-joins) that is not the FID. FIDs are not stable across file conversions that users might do."_

Haha, I wrote that comment in Coastal Vulnerability after Jess converted a GPKG to a SHP because Arc didn't have full-fledged GPKG support yet. And then her table-joins were off by one.

emlys reviewed (commented):
Looks good! Had a couple of questions. My only other suggestion would be to add a few comments pointing out the key parts of the algorithm.

richpsharp committed:
re #148 adding a unique ID to each feature and documenting some of the algorithm

richpsharp committed:
re #148 the test list needs to be a list too

richpsharp commented:
> Looks good! Had a couple of questions. My only other suggestion would be to add a few comments pointing out the key parts of the algorithm.

Sure, I added some more comments in 7c6d48beb86f5530211cf739b82866fea8baa37b. Do you think it's okay now?

richpsharp commented:
> Looks good! Had a couple of questions. My only other suggestion would be to add a few comments pointing out the key parts of the algorithm.

Thanks @emlys, I addressed all your comments and requested a re-review.

richpsharp committed:
re #148 better docstring for the fields

emlys commented:
Thanks @richpsharp !
